File: jsthegoodparts/memoize.js
function memoize(memo, formula) {
    return function ret(n) {
        var val = memo[n];
        if (typeof val === 'undefined') {
            val = formula(ret, n);
            memo[n] = val;
            console.log(memo);
        }
        return val;
    };
}

var fibonacci = memoize({0:0, 1:1}, function(f, n) {return f(n-1) + f(n-2);});
var factorial = memoize({0:0, 1:1}, function(f, n) {return f(n-1) * n;});

console.log(fibonacci(10));
console.log(factorial(10));

File: jsthegoodparts/memoize.py
#!/usr/bin/env python

def memoize(memo, formula):
    def ret(n):
        if n not in memo:
            memo[n] = formula(ret, n)
        return memo[n]
    return ret

fibonacci = memoize({0:0, 1:1}, lambda f, n: f(n-1) + f(n-2))
factorial = memoize({1:1}, lambda f, n: f(n-1) * n)

print fibonacci(10)
print factorial(10)
